### PR TITLE
Refactored the repository

### PR DESCRIPTION
**Pull request by abidahmadq**

### CODE REVIEW

**Comment by abidahmadq** [comment]
## PR Reviewer Guide 🔍

Here are some key observations to aid the review process:

<table>
<tr><td>⏱️&nbsp;<strong>Estimated effort to review</strong>: 3 🔵🔵🔵⚪⚪</td></tr>
<tr><td>🧪&nbsp;<strong>No relevant tests</strong></td></tr>
<tr><td>🔒&nbsp;<strong>No security concerns identified</strong></td></tr>
<tr><td>⚡&nbsp;<strong>Recommended focus areas for review</strong><br><br>

<details><summary><a href='https://github.com/abidahmadq/IndianFoodImageClassificationAPI/pull/43/files#diff-d6af0459a37d985953d7040c14f53feb3b9cc9e58b543aa3c2b80256d276c5e0R49-R63'><strong>TensorFlow Session Use</strong></a>

Using explicit tf.Session and tf.get_default_graph suggests TF1-style execution; with tensorflow==2.4.1 this can be fragile. Validate that the environment enables TF1 compatibility (e.g., disabling eager execution) and that concurrency in Flask won’t cause graph/session race conditions.
</summary>

```python
def __init__(self, model_path):
    self.sess = tf.Session()
    self.graph = tf.get_default_graph()
    with self.sess.as_default():
        with self.graph.as_default():
            self.model = load_model(model_path)
    logger.info("Model loaded successfully")

def predict(self, image_tensor):
    """Make prediction using the loaded model"""
    with self.sess.as_default():
        with self.graph.as_default():
            predictions = self.model.predict(image_tensor)
            predicted_class = np.argmax(predictions[0])
            return predicted_class, predictions[0]
```

</details>

<details><summary><a href='https://github.com/abidahmadq/IndianFoodImageClassificationAPI/pull/43/files#diff-d6af0459a37d985953d7040c14f53feb3b9cc9e58b543aa3c2b80256d276c5e0R98-R126'><strong>File Cleanup Robustness</strong></a>

Temporary files are removed after processing but not guaranteed on early failures; consider using try/finally blocks to ensure cleanup even when prediction or response construction fails.
</summary>

```python
try:
    # Save file
    filename = secure_filename(file.filename)
    filepath = os.path.join(app.config['UPLOAD_FOLDER'], filename)
    file.save(filepath)

    # Process image and predict
    image_tensor = load_image(filepath)
    predicted_class, probabilities = model_service.predict(image_tensor)

    # Create response
    response = {
        "predicted_class": int(predicted_class),
        "predicted_food": FOOD_CLASSES[predicted_class],
        "probabilities": {
            food_class: float(prob)
            for food_class, prob in zip(FOOD_CLASSES, probabilities)
        }
    }

    # Cleanup
    os.remove(filepath)

    return jsonify(response), 200

except Exception as e:
    logger.error(f"Error processing request: {str(e)}")
    abort(500, description="Error processing image")

```

</details>

<details><summary><a href='https://github.com/abidahmadq/IndianFoodImageClassificationAPI/pull/43/files#diff-37a968442c22648ceb4a16069820cc5350b97e50f45f26c5442d7c036ad743d2R7-R47'><strong>Static Asset Paths</strong></a>

Static CSS/JS/images are referenced via relative paths; in Flask, url_for('static', ...) is more reliable to avoid broken links when routing or hosting context changes.
</summary>

```html
    <link rel="stylesheet" href="../static/css/style.css">
    <link href="https://fonts.googleapis.com/css2?family=Poppins:wght@300;400;500;600&display=swap" rel="stylesheet">
</head>
<body>
    <div class="container">
        <header>
            <h1>Indian Food Classifier</h1>
            <p>Upload images of Indian food to classify them!</p>
        </header>

        <main>
            <div class="upload-section">
                <div class="upload-box" id="dropZone">
                    <input type="file" id="fileInput" accept=".jpg,.jpeg,.png" multiple>
                    <div class="upload-content">
                        <img src="../static/images/upload-icon.svg" alt="Upload" class="upload-icon">
                        <p>Drag & Drop images here or <span class="browse-text">Browse</span></p>
                        <p class="small-text">Supported formats: JPG, JPEG, PNG</p>
                    </div>
                </div>
                <button id="predictBtn" class="predict-btn" disabled>Predict</button>
            </div>

            <div class="preview-section">
                <h2>Selected Images</h2>
                <div id="imagePreview" class="image-preview-container"></div>
            </div>

            <div class="results-section">
                <h2>Results</h2>
                <div id="results" class="results-container"></div>
            </div>
        </main>

        <div id="loadingOverlay" class="loading-overlay">
            <div class="spinner"></div>
            <p>Processing...</p>
        </div>
    </div>

    <script src="../static/js/app.js"></script>
```

</details>

</td></tr>
</table>

**Comment by abidahmadq** [comment]
## Generating PR code suggestions


Work in progress ...<br>

**Comment by abidahmadq** [comment]
## PR Code Suggestions ✨


Latest suggestions up to 864de43
<table><thead><tr><td><strong>Category</strong></td><td align=left><strong>Suggestion&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></td><td align=center><strong>Impact</strong></td></tr><tbody><tr><td rowspan=1>Possible issue</td>
<td>



<details><summary>Remove TF1 session/graph usage</summary>

___


**Creating a dedicated TensorFlow session and graph in TensorFlow 2.x can conflict <br>with eager execution and lead to runtime errors. Remove manual session/graph <br>management and rely on TF2/Keras default execution to avoid crashes and <br>thread-safety issues. Load the model directly and call it without custom session <br>contexts.**

[run.py [48-56]](https://github.com/abidahmadq/IndianFoodImageClassificationAPI/pull/43/files#diff-d6af0459a37d985953d7040c14f53feb3b9cc9e58b543aa3c2b80256d276c5e0R48-R56)

```diff
 class ModelService:
     def __init__(self, model_path):
-        self.sess = tf.Session()
-        self.graph = tf.get_default_graph()
-        with self.sess.as_default():
-            with self.graph.as_default():
-                self.model = load_model(model_path)
-    logger.info("Model loaded successfully")
+        self.model = load_model(model_path)
+        logger.info("Model loaded successfully")
 
+    def predict(self, image_tensor):
+        predictions = self.model.predict(image_tensor)
+        predicted_class = int(np.argmax(predictions[0]))
+        return predicted_class, predictions[0]
+
```
<details><summary>Suggestion importance[1-10]: 8</summary>

__

Why: The PR uses `tf.Session()` and `tf.get_default_graph()` which are TF1 patterns and can cause issues under TF2 eager execution; removing manual session/graph handling improves correctness and stability. The improved_code aligns with TF2/Keras usage and accurately reflects the suggested change.


</details></details></td><td align=center>Medium

</td></tr><tr><td rowspan=2>General</td>
<td>



<details><summary>Support streams in image loader</summary>

___


**The loader only accepts filesystem paths, but calling it with a file stream improves <br>safety and performance. Accept both file-like objects and paths to prevent type <br>errors when switching to in-memory processing.**

[run.py [36-46]](https://github.com/abidahmadq/IndianFoodImageClassificationAPI/pull/43/files#diff-d6af0459a37d985953d7040c14f53feb3b9cc9e58b543aa3c2b80256d276c5e0R36-R46)

```diff
-def load_image(img_path, target_size=(224, 224)):
-    """Load and preprocess image for prediction"""
+def load_image(source, target_size=(224, 224)):
+    """Load and preprocess image from path or file-like stream"""
     try:
-        img = image.load_img(img_path, target_size=target_size)
+        img = image.load_img(source, target_size=target_size)
         img_tensor = image.img_to_array(img)
         img_tensor = np.expand_dims(img_tensor, axis=0)
-        img_tensor /= 255.
+        img_tensor = img_tensor.astype("float32") / 255.0
         return img_tensor
     except Exception as e:
         logger.error(f"Error processing image: {str(e)}")
         raise
```
<details><summary>Suggestion importance[1-10]: 7</summary>

__

Why: Extending `load_image` to accept both paths and file-like objects enables safer in-memory processing and avoids temp files, matching Keras `load_img` capabilities. The improved_code accurately applies the change with minor normalization improvement.

</details></details></td><td align=center>Medium

</td></tr><tr><td>



<details><summary>Process uploads in-memory</summary>

___


**Saving uploads to disk and then deleting them risks leaving orphan files on <br>exceptions and is unnecessary. Process files in-memory using <br><code>image.load_img(file.stream, ...)</code> to avoid filesystem race conditions and cleanup <br>failures.**

[run.py [100-121]](https://github.com/abidahmadq/IndianFoodImageClassificationAPI/pull/43/files#diff-d6af0459a37d985953d7040c14f53feb3b9cc9e58b543aa3c2b80256d276c5e0R100-R121)

```diff
-filename = secure_filename(file.filename)
-filepath = os.path.join(app.config['UPLOAD_FOLDER'], filename)
-file.save(filepath)
+# Process image directly from stream without saving
+image_tensor = load_image(file.stream)
+predicted_class, probabilities = model_service.predict(image_tensor)
 ...
-os.remove(filepath)
```
<details><summary>Suggestion importance[1-10]: 6</summary>

__

Why: Processing files directly from the request stream avoids filesystem I/O and cleanup risks, improving robustness; however, it requires broader refactoring (loader accepting streams) and is not strictly necessary for correctness. The concept is sound but impacts multiple lines and routes.


</details></details></td><td align=center>Low

</td></tr></tr></tbody></table>

___

#### Previous suggestions
<details><summary>Suggestions up to commit 864de43</summary>
<br><table><thead><tr><td><strong>Category</strong></td><td align=left><strong>Suggestion&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></td><td align=center><strong>Impact</strong></td></tr><tbody><tr><td rowspan=2>Possible issue</td>
<td>



<details><summary>Always cleanup uploaded files</summary>

___


**Ensure temporary files are always removed even if prediction or response creation <br>fails. Use a try/finally with existence check to avoid leaving sensitive uploads on <br>disk.**

[run.py [100-121]](https://github.com/abidahmadq/IndianFoodImageClassificationAPI/pull/43/files#diff-d6af0459a37d985953d7040c14f53feb3b9cc9e58b543aa3c2b80256d276c5e0R100-R121)

```diff
-file.save(filepath)
+try:
+    file.save(filepath)
 
-# Process image and predict
-image_tensor = load_image(filepath)
-predicted_class, probabilities = model_service.predict(image_tensor)
+    image_tensor = load_image(filepath)
+    predicted_class, probabilities = model_service.predict(image_tensor)
 
-# Create response
-response = {
-    "predicted_class": int(predicted_class),
-    "predicted_food": FOOD_CLASSES[predicted_class],
-    "probabilities": {
-        food_class: float(prob)
-        for food_class, prob in zip(FOOD_CLASSES, probabilities)
+    response = {
+        "predicted_class": int(predicted_class),
+        "predicted_food": FOOD_CLASSES[predicted_class],
+        "probabilities": {
+            food_class: float(prob)
+            for food_class, prob in zip(FOOD_CLASSES, probabilities)
+        }
     }
-}
 
-# Cleanup
-os.remove(filepath)
-    
-return jsonify(response), 200
+    return jsonify(response), 200
+finally:
+    try:
+        if os.path.exists(filepath):
+            os.remove(filepath)
+    except Exception as cleanup_err:
+        logger.warning(f"Failed to remove temp file {filepath}: {cleanup_err}")
```
<details><summary>Suggestion importance[1-10]: 8</summary>

__

Why: Ensuring temp file removal via try/finally is correct and important to prevent leftover uploads on disk even on errors; the proposed code matches the context and improves robustness.


</details></details></td><td align=center>Medium

</td></tr><tr><td>



<details><summary>Validate prediction output shape</summary>

___


**Guard against invalid prediction outputs to avoid <code>IndexError</code>/<code>ValueError</code> when the <br>model returns unexpected shapes. Validate <code>predictions</code> shape before indexing and <br>raise a clear error if invalid.**

[run.py [48-64]](https://github.com/abidahmadq/IndianFoodImageClassificationAPI/pull/43/files#diff-d6af0459a37d985953d7040c14f53feb3b9cc9e58b543aa3c2b80256d276c5e0R48-R64)

```diff
 class ModelService:
     def __init__(self, model_path):
         self.sess = tf.Session()
         self.graph = tf.get_default_graph()
         with self.sess.as_default():
             with self.graph.as_default():
                 self.model = load_model(model_path)
         logger.info("Model loaded successfully")
 
     def predict(self, image_tensor):
         """Make prediction using the loaded model"""
         with self.sess.as_default():
             with self.graph.as_default():
                 predictions = self.model.predict(image_tensor)
-                predicted_class = np.argmax(predictions[0])
-                return predicted_class, predictions[0]
+                if predictions is None or len(predictions) == 0:
+                    raise RuntimeError("Empty prediction output")
+                probs = predictions[0]
+                if probs is None or np.ndim(probs) != 1 or probs.size != len(FOOD_CLASSES):
+                    raise RuntimeError(f"Unexpected prediction shape: {np.shape(probs)}")
+                predicted_class = int(np.argmax(probs))
+                return predicted_class, probs
```
<details><summary>Suggestion importance[1-10]: 6</summary>

__

Why: Adding validation for model outputs reduces risk of runtime errors and gives clearer failures; it's contextually correct but a minor safeguard rather than a critical fix.

</details></details></td><td align=center>Low

</td></tr><tr><td rowspan=1>Security</td>
<td>



<details><summary>Harden file name validation</summary>

___


**Add a filename length and path traversal guard to prevent edge cases where crafted <br>names bypass <code>secure_filename</code> expectations. Explicitly reject hidden files and names <br>that normalize outside the upload directory.**

[run.py [33-35]](https://github.com/abidahmadq/IndianFoodImageClassificationAPI/pull/43/files#diff-d6af0459a37d985953d7040c14f53feb3b9cc9e58b543aa3c2b80256d276c5e0R33-R35)

```diff
 def allowed_file(filename):
-    return '.' in filename and filename.rsplit('.', 1)[1].lower() in ALLOWED_EXTENSIONS
+    if not filename or filename.startswith('.'):
+        return False
+    if '.' not in filename:
+        return False
+    ext = filename.rsplit('.', 1)[1].lower()
+    if ext not in ALLOWED_EXTENSIONS:
+        return False
+    # basic path traversal guard
+    if os.path.sep in filename or os.path.altsep and os.path.altsep in filename:
+        return False
+    return True
```
<details><summary>Suggestion importance[1-10]: 7</summary>

__

Why: Strengthening `allowed_file` against hidden files and simple path traversal improves security and is accurate for the shown code; impact is moderate since `secure_filename` is also used.


</details></details></td><td align=center>Medium

</td></tr></tr></tbody></table>

</details>
